### PR TITLE
Paper improvements

### PR DESCRIPTION
**Pull request by ike709**
## About The Pull Request

Fixes the potential stamp exploit we found during the last merge-a-thon and does some other cleanup while I'm here. Look at the CL for specifics.

Survived basic local testing.

## Changelog
:cl:
fix: Paper with pre-written text will now have the correct icon state
fix: Fixed a theoretical paper exploit for injecting fake stamp data
fix: Guarantee that paper inputs over the max length will be truncated
refactor: Minor backend optimizations and improvements to paper
/:cl:

### CODE REVIEW

File: code/modules/paperwork/paper.dm [hunk 1]
```diff
@@ -298,56 +298,62 @@
...
 				// the generated html might get beyond this limit
-				log_paper("[key_name(ui.user)] writing to paper [name], and overwrote it by [paper_len-MAX_PAPER_LENGTH]")
-			if(paper_len == 0)
-				to_chat(ui.user, pick("Writing block strikes again!", "You forgot to write anthing!"))
+				log_paper("[key_name(ui.user)] writing to paper [name], and overwrote it by [paper_len - MAX_PAPER_LENGTH] characters (this may be due to internal HTML)")
+				in_paper = copytext(in_paper, 1, MAX_PAPER_LENGTH)
+				to_chat(ui.user, "You run out of room on the paper!")
 			else
 				log_paper("[key_name(ui.user)] writing to paper [name]")
-				if(info != in_paper)
```
Comment (ike709) [inline code/modules/paperwork/paper.dm:343]: This check was removed under the assumption that sending an unnecessary TGUI update is cheaper than doing potentially really long string comparisons.

File: code/modules/paperwork/paper.dm [hunk 2]
```diff
@@ -298,56 +298,62 @@
...
 			var/stamp_y = text2num(params["y"])
 			var/stamp_r = text2num(params["r"])	// rotation in degrees
-			var/stamp_icon_state = params["stamp_icon_state"]
-			var/stamp_class = params["stamp_class"]
+
+			var/obj/holding = ui.user.get_active_held_item()
+			var/real_stamp = istype(holding, /obj/item/stamp)
+			var/datum/asset/spritesheet/sheet = get_asset_datum(/datum/asset/spritesheet/simple/paper)
+			var/stamp_icon_state = real_stamp ? holding.icon_state : "FAKE"
+			var/stamp_class = real_stamp ? sheet.icon_class_name(holding.icon_state) : "FAKE"
```
Comment (KubeRoot) [inline code/modules/paperwork/paper.dm]: Why isn't this an early return? What does stamping with "FAKE" imply?

File: code/modules/paperwork/paper.dm [hunk 3]
```diff
@@ -298,56 +298,62 @@
 		return
 	switch(action)
 		if("stamp")
+			if(length(stamps) >= MAX_PAPER_STAMPS)
+				to_chat(usr, pick("You try to stamp but you miss!", "There is no where else you can stamp!"))
```
Comment (KubeRoot) [inline code/modules/paperwork/paper.dm]:
```suggestion
				to_chat(usr, pick("You try to stamp but you miss!", "There is nowhere else you can stamp!"))
```

File: code/modules/paperwork/paper.dm [hunk 4]
```diff
@@ -298,56 +298,62 @@
...
+			update_static_data(usr,ui)
+			ui.user.visible_message("<span class='notice'>[ui.user] stamps [src] with [stamp_class]!</span>", "<span class='notice'>You stamp [src] with \the [holding.name]!</span>")
 
 		if("save")
 			var/in_paper = params["text"]
 			var/paper_len = length(in_paper)
 			field_counter = params["field_counter"] ? text2num(params["field_counter"]) : field_counter
 
+			if(paper_len == 0)
+				to_chat(ui.user, pick("Writing block strikes again!", "You forgot to write anthing!"))
```
Comment (KubeRoot) [inline code/modules/paperwork/paper.dm]:
```suggestion
				to_chat(ui.user, pick("Writing block strikes again!", "You forgot to write anything!"))
```

**Review by KubeRoot — CHANGES_REQUESTED** [review]

File: code/modules/paperwork/paper.dm (hunk 2 above)
Comment (ike709) [reply code/modules/paperwork/paper.dm]: `FAKE` results in using the fallback stamp icon, but there's no reason this can't just be an early return.

**Pull request merged**